content/en/platform/corda/5.0-dev-preview-2/developing/api/amqp-serialization-enum-evolution.md

```diff
@@ -0,0 +1,337 @@
+can transform the serialised object into the form as it exists in the deserializer. Newness is determined simply
+by length of the list of all transforms. This is sufficient as transform annotations should only ever be added
+
+
+{{< warning >}}
+technically there is nothing to prevent annotations being removed in newer versions. However,
+this will break backward compatibility and should thus be avoided unless a rigorous upgrade procedure
+is in place to cope with all deployed instances of the class and all serialised versions existing
+within vaults.
+{{< /warning >}}
+
+Thus, on deserialization, there will be two options to chose from in terms of transformation rules
+
+* Determined from the local class and the annotations applied to it (the local copy)
+* Parsed from the AMQP header (the remote copy)
+
+Which set is used will simply be the largest.
+
+## Renaming Constants
+
+Renamed constants are marked as such with the `@CordaSerializationTransformRenames` meta annotation that
+wraps a list of `@CordaSerializationTransformRename` annotations. Each rename requiring an instance in the
+list.
+
+Each instance must provide the new name of the constant as well as the old. For example, consider the following enumeration:
+
+```kotlin
+enum class Example {
+    A, B, C
+}
+```
+
+If we were to rename constant C to D this would be done as follows:
+
+```kotlin
+@CordaSerializationTransformRenames (
+    CordaSerializationTransformRename("D", "C")
+)
+enum class Example {
+    A, B, D
+}
+```
+
+{{< note >}}
+The parameters to the `CordaSerializationTransformRename` annotation are defined as ‘to’ and ‘from’,
+so in the above example it can be read as constant D (given that is how the class now exists) was renamed
+from C
+
+{{< /note >}}
+In the case where a single rename has been applied the meta annotation may be omitted. Thus, the following is
+functionally identical to the above:
+
+```kotlin
+@CordaSerializationTransformRename("D", "C")
+enum class Example {
+    A, B, D
+}
+```
+
+However, as soon as a second rename is made the meta annotation must be used. For example, if at some time later
+B is renamed to E:
+
+```kotlin
+@CordaSerializationTransformRenames (
+    CordaSerializationTransformRename(from = "B", to = "E"),
+    CordaSerializationTransformRename(from = "C", to = "D")
+)
+enum class Example {
+    A, E, D
+}
+```
+
+### Rules
+
+* A constant cannot be renamed to match an existing constant, this is enforced through language constraints
+* A constant cannot be renamed to a value that matches any previous name of any other constant
+
+If either of these covenants are inadvertently broken, a `NotSerializableException` will be thrown on detection
+by the serialization engine as soon as they are detected. Normally this will be the first time an object doing
+so is serialized. However, in some circumstances, it could be at the point of deserialization.
```

Comment:
doing what?
Maybe:
Usually, this occurs the first time the object is serialized.

Reply:
Done